// @mizchi/optools/lib/generator.mjs
import ts from "typescript";
import { rollup } from "rollup";
import dts from "rollup-plugin-dts";
export function emitLibDts(files, outDir, options) {
    const shouldOverrideConfigs = {
        outDir,
        declaration: true,
        emitDeclarationOnly: true,
        noEmit: true,
    };
    const host = ts.createCompilerHost({
        ...options,
        ...shouldOverrideConfigs,
    });
    const program = ts.createProgram(files, options, host);
    const emitResult = program.emit();
    return emitResult;
}
export async function generateBundleDts({ input, external, compilerOptions, respectExternal = false }) {
    const bundle = await rollup({
        input: input,
        external,
        plugins: [dts({
                respectExternal,
                compilerOptions: compilerOptions,
            })],
    });
    const out = await bundle.generate({
        format: "es",
    });
    const dtsCode = out.output.find((o) => {
        if (o.type === "chunk" && o.fileName.endsWith(".d.ts")) {
            return true;
        }
    });
    return dtsCode.code;
}
